File: scripts/lib/external-mod-runtimes.mjs
import { existsSync } from "node:fs";
import { readFile } from "node:fs/promises";
import os from "node:os";
import path from "node:path";
import { fileURLToPath } from "node:url";

const REPO_ROOT = fileURLToPath(new URL("../../", import.meta.url));
const DEFAULT_DESCRIPTOR_URL = new URL(
  "../../config/external-mod-runtimes.json",
  import.meta.url,
);
const ID_PATTERN = /^[a-z][a-z0-9-]{2,63}$/;
const ENV_PATTERN = /^(?:NEWMA_DOCK|VIBEDESK)_[A-Z0-9_]+$/;

function objectValue(value, label) {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw new Error(`${label} must be an object`);
  }
  return value;
}

function stringValue(value, label) {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${label} must be a non-empty string`);
  }
  return value;
}

function envName(value, label) {
  const name = stringValue(value, label);
  if (!ENV_PATTERN.test(name)) {
    throw new Error(`${label} must be a NEWMA_DOCK_* or VIBEDESK_* name`);
  }
  return name;
}

function configuredEnvValue(env, name) {
  const configured = env[name]?.trim();
  if (configured || !name.startsWith("NEWMA_DOCK_")) return configured;
  return env[`VIBEDESK_${name.slice("NEWMA_DOCK_".length)}`]?.trim();
}

function safeRelativePath(value, label) {
  const candidate = stringValue(value, label);
  if (
    path.isAbsolute(candidate) ||
    candidate.includes("\\") ||
    candidate.split("/").some((part) => !part || part === "." || part === "..")
  ) {
    throw new Error(`${label} must be a safe relative path`);
  }
  return candidate;
}

function exactHttpOrigin(value, label) {
  let parsed;
  try {
    parsed = new URL(value);
  } catch {
    throw new Error(`${label} must be an HTTP(S) origin`);
  }
  if (
    !["http:", "https:"].includes(parsed.protocol) ||
    parsed.username ||
    parsed.password ||
    parsed.origin !== value
  ) {
    throw new Error(`${label} must be an HTTP(S) origin`);
  }
  return parsed;
}

function endpointPort(url) {
  if (url.port) return Number(url.port);
  return url.protocol === "https:" ? 443 : 80;
}

function localHostname(hostname) {
  return ["127.0.0.1", "localhost", "::1", "[::1]"].includes(hostname);
}

function resolveRoots(rawRoots, { repoRoot, homeDir, env }) {
  const roots = { repo: repoRoot };
  if (!Array.isArray(rawRoots)) throw new Error("descriptor.roots must be an array");
  for (const [index, rawRoot] of rawRoots.entries()) {
    const root = objectValue(rawRoot, `descriptor.roots[${index}]`);
    const id = stringValue(root.id, `descriptor.roots[${index}].id`);
    if (!ID_PATTERN.test(id) || id === "repo" || Object.hasOwn(roots, id)) {
      throw new Error(`descriptor.roots[${index}].id is invalid or duplicated`);
    }
    const configured = configuredEnvValue(
      env,
      envName(root.env, `descriptor.roots[${index}].env`),
    );
    if (configured) {
      roots[id] = path.resolve(repoRoot, configured);
      continue;
    }
    const fallback = objectValue(root.fallback, `descriptor.roots[${index}].fallback`);
    const fallbackPath = stringValue(
      fallback.path,
      `descriptor.roots[${index}].fallback.path`,
    );
    if (fallback.type === "repo-relative") {
      roots[id] = path.resolve(repoRoot, fallbackPath);
    } else if (fallback.type === "home-relative") {
      roots[id] = path.resolve(homeDir, fallbackPath);
    } else {
      throw new Error(`descriptor.roots[${index}].fallback.type is unsupported`);
    }
  }
  return roots;
}

function resolveWorkspace(raw, label, { roots, repoRoot, env, exists }) {
  const workspace = objectValue(raw, label);
  const configuredEnv = envName(workspace.env, `${label}.env`);
  const configured = configuredEnvValue(env, configuredEnv);
  if (configured) {
    const resolved = path.resolve(repoRoot, configured);
    return {
      path: exists(resolved) ? resolved : null,
      source: exists(resolved) ? "environment" : "missing-environment",
      env: configuredEnv,
      attempts: [resolved],
    };
  }
  if (!Array.isArray(workspace.candidates) || workspace.candidates.length === 0) {
    throw new Error(`${label}.candidates must be a non-empty array`);
  }
  const attempts = [];
  for (const [index, rawCandidate] of workspace.candidates.entries()) {
    const candidate = objectValue(rawCandidate, `${label}.candidates[${index}]`);
    const rootId = stringValue(candidate.root, `${label}.candidates[${index}].root`);
    const root = roots[rootId];
    if (!root) throw new Error(`${label}.candidates[${index}].root is unknown`);
    const relative = safeRelativePath(
      candidate.path,
      `${label}.candidates[${index}].path`,
    );
    const resolved = path.resolve(root, relative);
    attempts.push(resolved);
    if (exists(resolved)) {
      return {
        path: resolved,
        source: "discovered",
        env: configuredEnv,
        attempts,
      };
    }
  }
  return { path: null, source: "missing", env: configuredEnv, attempts };
}

function resolveEndpoint(raw, label, env) {
  const endpoint = objectValue(raw, label);
  const configuredEnv = envName(endpoint.env, `${label}.env`);
  const parsed = exactHttpOrigin(
    configuredEnvValue(env, configuredEnv) ||
      stringValue(endpoint.defaultOrigin, `${label}.defaultOrigin`),
    `${label}.origin`,
  );
  const healthPath = stringValue(endpoint.healthPath, `${label}.healthPath`);
  if (!healthPath.startsWith("/") || healthPath.startsWith("//") || healthPath.includes("..")) {
    throw new Error(`${label}.healthPath must be a safe absolute path`);
  }
  return {
    env: configuredEnv,
    origin: parsed.origin,
    port: endpointPort(parsed),
    local: localHostname(parsed.hostname),
    healthPath,
    healthUrl: new URL(healthPath, `${parsed.origin}/`).toString(),
  };
}

export function resolveExternalModRuntimes(
  descriptor,
  {
    repoRoot = REPO_ROOT,
    homeDir = os.homedir(),
    env = process.env,
    exists = existsSync,
  } = {},
) {
  const value = objectValue(descriptor, "descriptor");
  if (value.schemaVersion !== "1.0") throw new Error("unsupported runtime descriptor version");
  const roots = resolveRoots(value.roots, { repoRoot, homeDir, env });
  if (!Array.isArray(value.runtimes) || value.runtimes.length === 0) {
    throw new Error("descriptor.runtimes must be a non-empty array");
  }
  const runtimes = value.runtimes.map((rawRuntime, index) => {
    const runtime = objectValue(rawRuntime, `descriptor.runtimes[${index}]`);
    const id = stringValue(runtime.id, `descriptor.runtimes[${index}].id`);
    if (!ID_PATTERN.test(id)) throw new Error(`${id} is not a valid runtime id`);
    const workspaces = objectValue(runtime.workspaces, `${id}.workspaces`);
    const endpoints = objectValue(runtime.endpoints, `${id}.endpoints`);
    return {
      id,
      label: stringValue(runtime.label, `${id}.label`),
      adapter: stringValue(runtime.adapter, `${id}.adapter`),
      workspaces: Object.fromEntries(
        Object.entries(workspaces).map(([name, workspace]) => [
          name,
          resolveWorkspace(workspace, `${id}.workspaces.${name}`, {
            roots,
            repoRoot,
            env,
            exists,
          }),
        ]),
      ),
      endpoints: Object.fromEntries(
        Object.entries(endpoints).map(([name, endpoint]) => [
          name,
          resolveEndpoint(endpoint, `${id}.endpoints.${name}`, env),
        ]),
      ),
    };
  });
  const ids = runtimes.map(({ id }) => id);
  if (new Set(ids).size !== ids.length) throw new Error("runtime ids must be unique");
  return {
    schemaVersion: "1.0",
    roots,
    runtimes,
    byId: Object.fromEntries(runtimes.map((runtime) => [runtime.id, runtime])),
  };
}

export async function loadExternalModRuntimes({
  descriptorUrl = DEFAULT_DESCRIPTOR_URL,
  ...options
} = {}) {
  const descriptor = JSON.parse(await readFile(descriptorUrl, "utf8"));
  return resolveExternalModRuntimes(descriptor, options);
}

export function runtimeEnvironment(catalog) {
  const result = {};
  for (const runtime of catalog.runtimes) {
    for (const workspace of Object.values(runtime.workspaces)) {
      if (workspace.path) result[workspace.env] = workspace.path;
    }
    for (const endpoint of Object.values(runtime.endpoints)) {
      result[endpoint.env] = endpoint.origin;
    }
  }
  return result;
}

File: services/api/vibe_visualization_api/external_mod_runtimes.py
import json
import os
from functools import lru_cache
from pathlib import Path
from typing import Any, Mapping
from urllib.parse import urlsplit


def _default_project_root() -> Path:
    return Path(__file__).resolve().parents[3]


def _default_descriptor_path() -> Path:
    return _default_project_root() / "config" / "external-mod-runtimes.json"


@lru_cache(maxsize=4)
def load_runtime_descriptor(
    descriptor_path: Path | None = None,
) -> dict[str, Any]:
    path = (descriptor_path or _default_descriptor_path()).resolve()
    value = json.loads(path.read_text(encoding="utf-8"))
    if not isinstance(value, dict) or value.get("schemaVersion") != "1.0":
        raise ValueError("unsupported external Mod runtime descriptor")
    if not isinstance(value.get("roots"), list) or not isinstance(
        value.get("runtimes"), list
    ):
        raise ValueError("invalid external Mod runtime descriptor")
    return value


def _runtime_by_id(descriptor: Mapping[str, Any], runtime_id: str) -> Mapping[str, Any]:
    for runtime in descriptor["runtimes"]:
        if isinstance(runtime, dict) and runtime.get("id") == runtime_id:
            return runtime
    raise KeyError(f"unknown external Mod runtime: {runtime_id}")


def _configured_env(env: Mapping[str, str], name: str) -> str:
    prefixes = ("NEWMA_DESK_", "NEWMA_DOCK_", "VIBEDESK_")
    suffix = next(
        (name.removeprefix(prefix) for prefix in prefixes if name.startswith(prefix)),
        None,
    )
    if suffix is None:
        return env.get(name, "").strip()
    for prefix in prefixes:
        configured = env.get(f"{prefix}{suffix}", "").strip()
        if configured:
            return configured
    return ""


def _resolve_roots(
    descriptor: Mapping[str, Any],
    *,
    project_root: Path,
    home: Path,
    env: Mapping[str, str],
) -> dict[str, Path]:
    roots = {"repo": project_root.resolve()}
    for root in descriptor["roots"]:
        if not isinstance(root, dict):
            raise ValueError("runtime root must be an object")
        root_id = str(root["id"])
        configured = _configured_env(env, str(root["env"]))
        if configured:
            path = Path(configured).expanduser()
            roots[root_id] = (
                path if path.is_absolute() else project_root / path
            ).resolve()
            continue
        fallback = root.get("fallback")
        if not isinstance(fallback, dict):
            raise ValueError("runtime root fallback must be an object")
        relative = Path(str(fallback["path"])).expanduser()
        if fallback.get("type") == "repo-relative":
            roots[root_id] = (project_root / relative).resolve()
        elif fallback.get("type") == "home-relative":
            roots[root_id] = (home / relative).resolve()
        else:
            raise ValueError("unsupported runtime root fallback")
    return roots


def resolve_runtime_workspace(
    runtime_id: str,
    workspace_name: str,
    *,
    descriptor_path: Path | None = None,
    project_root: Path | None = None,
    home: Path | None = None,
    env: Mapping[str, str] | None = None,
) -> Path:
    descriptor = load_runtime_descriptor(descriptor_path)
    runtime = _runtime_by_id(descriptor, runtime_id)
    workspaces = runtime.get("workspaces")
    if not isinstance(workspaces, dict) or workspace_name not in workspaces:
        raise KeyError(f"unknown {runtime_id} workspace: {workspace_name}")
    workspace = workspaces[workspace_name]
    if not isinstance(workspace, dict):
        raise ValueError("runtime workspace must be an object")
    root = (project_root or _default_project_root()).resolve()
    environment = os.environ if env is None else env
    configured = _configured_env(environment, str(workspace["env"]))
    if configured:
        path = Path(configured).expanduser()
        return (path if path.is_absolute() else root / path).resolve()
    roots = _resolve_roots(
        descriptor,
        project_root=root,
        home=(home or Path.home()).resolve(),
        env=environment,
    )
    candidates = workspace.get("candidates")
    if not isinstance(candidates, list) or not candidates:
        raise ValueError("runtime workspace candidates are required")
    resolved: list[Path] = []
    for candidate in candidates:
        if not isinstance(candidate, dict):
            raise ValueError("runtime workspace candidate must be an object")
        root_path = roots.get(str(candidate["root"]))
        if root_path is None:
            raise ValueError("runtime workspace candidate references an unknown root")
        path = (root_path / str(candidate["path"])).resolve()
        resolved.append(path)
        if path.exists():
            return path
    return resolved[0]


def resolve_runtime_origin(
    runtime_id: str,
    endpoint_name: str,
    *,
    descriptor_path: Path | None = None,
    env: Mapping[str, str] | None = None,
) -> str:
    descriptor = load_runtime_descriptor(descriptor_path)
    runtime = _runtime_by_id(descriptor, runtime_id)
    endpoints = runtime.get("endpoints")
    if not isinstance(endpoints, dict) or endpoint_name not in endpoints:
        raise KeyError(f"unknown {runtime_id} endpoint: {endpoint_name}")
    endpoint = endpoints[endpoint_name]
    if not isinstance(endpoint, dict):
        raise ValueError("runtime endpoint must be an object")
    environment = os.environ if env is None else env
    value = _configured_env(environment, str(endpoint["env"])) or str(
        endpoint["defaultOrigin"]
    )
    parsed = urlsplit(value)
    if (
        parsed.scheme not in {"http", "https"}
        or not parsed.netloc
        or parsed.username is not None
        or parsed.password is not None
        or parsed.path not in {"", "/"}
        or parsed.query
        or parsed.fragment
    ):
        raise ValueError("runtime endpoint must be an HTTP(S) origin")
    return f"{parsed.scheme}://{parsed.netloc}"


def default_external_origins(
    *,
    descriptor_path: Path | None = None,
    env: Mapping[str, str] | None = None,
) -> list[str]:
    descriptor = load_runtime_descriptor(descriptor_path)
    origins: list[str] = []
    for runtime in descriptor["runtimes"]:
        for endpoint_name in runtime.get("endpoints", {}):
            origin = resolve_runtime_origin(
                str(runtime["id"]),
                str(endpoint_name),
                descriptor_path=descriptor_path,
                env=env,
            )
            if origin not in origins:
                origins.append(origin)
    return origins
